src/instrumentation/podmutator.go

```diff
@@ -35,150 +45,250 @@ var (
+	getLanguageInstrumentationsFn func(insts []*v1alpha2.Instrumentation) ([]*v1alpha2.Instrumentation, error)
+	logger                        logr.Logger
+	operatorNamespace             string
 }
 
-var _ webhookhandler.PodMutator = (*instPodMutator)(nil)
-
-func NewMutator(logger logr.Logger, client client.Client) *instPodMutator {
+// NewMutator is used to get a new instance of a mutator
+func NewMutator(
+	logger logr.Logger,
+	client client.Client,
+	sdkInjector SdkInjector,
+	secretReplicator SecretReplicator,
+	instrumentationLocator InstrumentationLocator,
+	getLicenseKeySecretFn func(insts []*v1alpha2.Instrumentation) (string, error),
+	getLanguageInstrumentationsFn func(insts []*v1alpha2.Instrumentation) ([]*v1alpha2.Instrumentation, error),
+	operatorNamespace string,
+) *instPodMutator {
 	return &instPodMutator{
-		Logger: logger,
-		Client: client,
-		sdkInjector: &sdkInjector{
-			logger: logger,
-			client: client,
-		},
+		logger:                        logger,
+		client:                        client,
+		instrumentationLocator:        instrumentationLocator,
+		sdkInjector:                   sdkInjector,
+		secretReplicator:              secretReplicator,
+		getLicenseKeySecretFn:         getLicenseKeySecretFn,
+		getLanguageInstrumentationsFn: getLanguageInstrumentationsFn,
+		operatorNamespace:             operatorNamespace,
 	}
 }
 
+// Mutate is used to mutate a pod based on some instrumentation(s)
 func (pm *instPodMutator) Mutate(ctx context.Context, ns corev1.Namespace, pod corev1.Pod) (corev1.Pod, error) {
-	logger := pm.Logger.WithValues("namespace", pod.Namespace, "name", pod.Name)
-
-	var inst *v1alpha1.Instrumentation
-	var err error
-
-	insts := languageInstrumentations{}
+	logger := pm.logger.WithValues("namespace", pod.Namespace, "name", pod.Name)
 
-	// We bail out if any annotation fails to process.
-
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectJava); err != nil {
+	instCandidates, err := pm.instrumentationLocator.GetInstrumentations(ctx, ns, pod)
+	if err != nil {
 		// we still allow the pod to be created, but we log a message to the operator's logs
 		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
 		return pod, err
 	}
-	insts.Java = inst
+	if len(instCandidates) == 0 {
+		logger.Info("no New Relic Instrumentation instance for this Pod")
+		return pod, errNoInstancesAvailable
+	}
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectNodeJS); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+	instrumentations, err := pm.getLanguageInstrumentationsFn(instCandidates)
+	if err != nil {
+		if errors.Is(err, errMultipleInstancesPossible) {
+			pm.logger.Info("too many New Relic Instrumentation instances for this Pod.  only 1 allowed")
+			return pod, nil
```

Comment:
So in this case, we are not injecting/mutating anything and silently returning unmutated pod with no error. I think we should log this as an error `pm.logger.Error` and return error otherwise how would the upstream debug this scenario. Thoughts welcome.

Reply:
I need to retest this scenario

Reply:
updated this